### PR TITLE
CCORE-671-ES-Refactor Builds

### PR DESCRIPTION
**Pull request by evensolberg-ffatg**
### Purpose of this PR

Consolidate the build functionality into one function, which makes it easier to maintain and troubleshoot. A few of the "supporting" functions (most notably `UploadToAppCenter`) have been updated and made more robust as a consequence.

### PR Merge order

- This PR only

### Testing status

The Story and Showroom pipelines have been tested with this final version with the DEBUG flag on and off.

### Documentation status

In addition to an updated README, there is also a corresponding file to the `buildOnPlatform.groovy` file in the /docs/var/ folder that describes the new function.

### Technical risk

If the build function doesn't work, the pipeline doesn't build any Unity projects. Which is widely regarded as a Bad Thing (tm).

### CODE REVIEW

**Comment by bdebrincunity** [comment]
This is great, I like the docs and debug info approach.

Only thing I wonder is if `vars/buildOnPlatform.groovy` is doing a little too much now? To me it might more sense if just did the build, and post build steps (like archiving or UploadToAppCenter) A) remained in the main unityBuildPipeline (for now) or B) went into a separate groovy script file.

File: vars/buildOnPlatform.groovy [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+                } // if (DEBUG)
+                // Update permissions to OSX project, needed to launch
+                try { // making the binary an executable
+                    sh "chmod +x ${project}/${executable}.app/Contents/MacOS/${executable}"
+                } catch (errorMessage) {
+                    error 'ERROR: Unable set `chmod`. Error message: ' + errorMessage
+                } // try making the binary an executable
+            } // if ("${type}" == 'mac')
+
+            if (isTargetBranch) {
```
Comment (jrowinski3d) [inline vars/buildOnPlatform.groovy:101]: I don't believe this is working as the current **isTargetBranch** is set to 
`Boolean isTargetBranch = "${BRANCH_NAME}".matches('main|staging|develop|release')`

yet you have build artifacts zipped up with no VERSION attached to it, instead you have a trailing **-** 
(builder-macos-.zip)
(builder-windows64-.zip

Comment (jrowinski3d) [reply vars/buildOnPlatform.groovy:101]: The way I got it to work in **develop** on DockerPush() is: 

`if ("${env.isTargetBranch}".toBoolean())`

Comment (jrowinski3d) [reply vars/buildOnPlatform.groovy:101]: In theory, you shouldn't have even been able to send builds to AppCenter yet they were in a PR. A bit strange though is that the AppCenter build versioning seems to be fine. Guessing due to line 14 in UploadToAppCenter

`String buildVer = System.getenv('VERSION') ?: '0.1.0'`

**Comment by evensolberg-ffatg** [comment]
> This is great, I like the docs and debug info approach.
> 
> Only thing I wonder is if `vars/buildOnPlatform.groovy` is doing a little too much now? To me it might more sense if just did the build, and post build steps (like archiving or UploadToAppCenter) A) remained in the main unityBuildPipeline (for now) or B) went into a separate groovy script file.

Yeah, we could potentially refactor the refactor. :)
At least for now it makes the pipeline very clean. But you have a point.

**Review by jrowinski3d — CHANGES_REQUESTED** [review]

File: vars/UploadToAppCenter.groovy [hunk 2]
```diff
@@ -1,36 +1,70 @@
...
+
+String call(Map args = [:]) {
     Map argsDefault = [
-            projectType: "",
-            projectPath: "",
-            distGroups: "",
+            projectType: '',
+            projectPath: '',
+            distGroups: '',
+            DEBUG: ''
```
[the remark this answers is not in the document]
Comment (evensolberg-ffatg) [reply vars/UploadToAppCenter.groovy:8]: Hmm. Maybe, yes.

File: vars/buildOnPlatform.groovy (hunk 1 above)
Comment (evensolberg-ffatg) [reply vars/buildOnPlatform.groovy:101]: > I don't believe this is working as the current **isTargetBranch** is set to
> `Boolean isTargetBranch = "${BRANCH_NAME}".matches('main|staging|develop|release')`
> 
> yet you have build artifacts zipped up with no VERSION attached to it, instead you have a trailing **-**
> (builder-macos-.zip)
> (builder-windows64-.zip

Yeah, I agree. Will look into it.

**Review by bdebrincunity — APPROVED** [review]
Looks good! 👍

**Pull request merged**